File: instrumentation/src/test/java/io/opentelemetry/rum/internal/instrumentation/activity/RumFragmentActivityRegistererTest.java
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.rum.internal.instrumentation.activity;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.app.Activity;
import android.app.Application;
import androidx.fragment.app.FragmentActivity;
import androidx.fragment.app.FragmentManager;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class RumFragmentActivityRegistererTest {

    @Mock FragmentManager.FragmentLifecycleCallbacks fragmentCallbacks;

    @Test
    void createHappyPath() {
        FragmentActivity activity = mock(FragmentActivity.class);
        FragmentManager manager = mock(FragmentManager.class);

        when(activity.getSupportFragmentManager()).thenReturn(manager);

        Application.ActivityLifecycleCallbacks underTest =
                RumFragmentActivityRegisterer.create(fragmentCallbacks);

        underTest.onActivityPreCreated(activity, null);
        verify(manager).registerFragmentLifecycleCallbacks(fragmentCallbacks, true);
    }

    @Test
    void callbackIgnoresNonFragmentActivity() {
        Activity activity = mock(Activity.class);

        Application.ActivityLifecycleCallbacks underTest =
                RumFragmentActivityRegisterer.create(fragmentCallbacks);

        underTest.onActivityPreCreated(activity, null);
    }

    @Test
    void createPre29HappyPath() {
        FragmentActivity activity = mock(FragmentActivity.class);
        FragmentManager manager = mock(FragmentManager.class);

        when(activity.getSupportFragmentManager()).thenReturn(manager);

        Application.ActivityLifecycleCallbacks underTest =
                RumFragmentActivityRegisterer.createPre29(fragmentCallbacks);

        underTest.onActivityCreated(activity, null);
        verify(manager).registerFragmentLifecycleCallbacks(fragmentCallbacks, true);
    }

    @Test
    void pre29CallbackIgnoresNonFragmentActivity() {
        Activity activity = mock(Activity.class);

        Application.ActivityLifecycleCallbacks underTest =
                RumFragmentActivityRegisterer.createPre29(fragmentCallbacks);

        underTest.onActivityCreated(activity, null);
    }
}

File: services/src/test/java/io/opentelemetry/android/internal/services/visiblescreen/fragments/RumFragmentActivityRegistererTest.kt
/*
 * Copyright The OpenTelemetry Authors
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.internal.services.visiblescreen.fragments

import android.app.Activity
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.FragmentManager
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.mockk.junit5.MockKExtension
import io.mockk.mockk
import io.mockk.verify
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
internal class RumFragmentActivityRegistererTest {
    @MockK
    private lateinit var fragmentCallbacks: FragmentManager.FragmentLifecycleCallbacks

    @Test
    fun createHappyPath() {
        val activity = mockk<FragmentActivity>()
        val manager = mockk<FragmentManager>(relaxed = true)

        every { activity.supportFragmentManager } returns manager

        val underTest =
            RumFragmentActivityRegisterer.create(fragmentCallbacks)

        underTest.onActivityPreCreated(activity, null)
        verify {
            manager.registerFragmentLifecycleCallbacks(
                fragmentCallbacks,
                true,
            )
        }
    }

    @Test
    fun callbackIgnoresNonFragmentActivity() {
        val activity = mockk<Activity>()

        val underTest =
            RumFragmentActivityRegisterer.create(fragmentCallbacks)

        underTest.onActivityPreCreated(activity, null)
    }

    @Test
    fun createPre29HappyPath() {
        val activity = mockk<FragmentActivity>()
        val manager = mockk<FragmentManager>(relaxed = true)

        every { activity.supportFragmentManager } returns manager

        val underTest =
            RumFragmentActivityRegisterer.createPre29(fragmentCallbacks)

        underTest.onActivityCreated(activity, null)
        verify {
            manager.registerFragmentLifecycleCallbacks(
                fragmentCallbacks,
                true,
            )
        }
    }

    @Test
    fun pre29CallbackIgnoresNonFragmentActivity() {
        val activity = mockk<Activity>()

        val underTest =
            RumFragmentActivityRegisterer.createPre29(fragmentCallbacks)

        underTest.onActivityCreated(activity, null)
    }
}
